docs/sphinx/using/cudaq/platform.rst

```diff
@@ -45,9 +45,217 @@ Here is a simple example demonstrating this
 
     // Get the results, note future::get() will kick off a wait
     // if the results are not yet available.
-    for (auto& counts : countsFutures) {
+    for (auto& counts : countFutures) {
       counts.get().dump();
     }
 
 CUDA Quantum exposes asynchronous versions of the default :code:`cudaq::` algorithmic
-primitive functions like :code:`sample` and :code:`observe`. 
\ No newline at end of file
+primitive functions like :code:`sample` and :code:`observe` (e.g., :code:`cudaq::sample_async` function in the above code snippet). 
+
+One can then specify the target multi-QPU architecture (:code:`nvidia-mqpu`) with the :code:`--target` flag:
+ 
+.. code-block:: console 
+
+    nvq++ simple.cpp -target nvidia-mqpu
+    ./a.out
+
+Depending on the number of GPUs available on the system, the :code:`nvidia-mqpu` platform will create the same number of virtual QPU instances.
+For example, on a system with 4 GPUs, the above code will distribute the four sampling tasks among those :code:`GPUEmulatedQPU` instances.
+
+The results might look like the following (4 different random samplings).
+
+.. code-block:: console 
+  
+    Number of QPUs: 4
+    { 10011:28 01100:28 ... }
+    { 10011:37 01100:25 ... }
+    { 10011:29 01100:25 ... }
+    { 10011:33 01100:30 ... }
+
+.. note:: 
+
+  By default, the :code:`nvidia-mqpu` platform will utilize all available GPUs (number of QPUs instances is equal to the number of GPUs).
+  To specify the number QPUs to be instantiated, one can set the :code:`CUDAQ_MQPU_NGPUS` environment variable.
+  For example, :code:`export CUDAQ_MQPU_NGPUS=2` to specify that only 2 QPUs (GPUs) are needed.
+
+
+An equivalent example in Python is as follows:
+
+.. code:: python
+
+    import cudaq
+
+    cudaq.set_target("nvidia-mqpu")
+    target = cudaq.get_target()
+    num_qpus = target.num_qpus()
+    print("Number of QPUs:", num_qpus)
+
+    kernel, runtime_param = cudaq.make_kernel(int)
+    qubits = kernel.qalloc(runtime_param)
+    # Place qubits in superposition state.
+    kernel.h(qubits)
+    # Measure.
+    kernel.mz(qubits)
+
+    count_futures = []
+    for qpu in range(num_qpus):
+        count_futures.append(cudaq.sample_async(kernel, 5, qpu_id=qpu))
+
+
+    for counts in count_futures:
+      print(counts.get())
+
+Asynchronous expectation value computations
++++++++++++++++++++++++++++++++++++++++++++
+
+One typical use case of the :code:`nvidia-mqpu` platform is to distribute the 
+expectation value computations of a multi-term Hamiltonian across multiple virtual QPUs (:code:`GPUEmulatedQPU`).
+
+Here is an example.
+
+.. code-block:: cpp 
```

Comment:
Hm, I know we have quite a few code blocks in the docs, but directly including them in this form is fragile (as seen by your correction of the typo in the sample above). We should move towards instead having a C++/Python file that we run as part of CI, and including that file in the docs instead. It is possible to include only certain lines in the file, see the [docs](https://devopstutodoc.readthedocs.io/en/stable/documentation/doc_generators/sphinx/rest_sphinx/code/literalinclude/literalinclude.html). Could you move this as well as the other larger code examples in platform.rst into its own file, and build that file as part of CI? I am not sure if we want to include all examples from the docs in the docker image, so it may make sense to create a separate folder (e.g. docs/sphinx/snippets) that we build and run like we do the examples folder, but don't copy into the docker file.

Reply:
I've fixed it as suggested:

- Move code blocks into their own files under `docs/sphinx/snippets` and use `literalinclude` directive to pull them in the docs.

- Add them to CI: extend the CMake script to handle variations in location, test configurations, etc.

Since the current CI runner won't be able to run these added tests yet (need GPUs), I validated them on gorby. All the tests in `docs/` are passing :-) 

```
root@d48b0968fd6d:/cuda-quantum/build# ctest -R nvqpp_
Test project /cuda-quantum/build
      Start 255: nvqpp_GHZ
 1/20 Test #255: nvqpp_GHZ .........................   Passed    2.93 sec
      Start 256: nvqpp_MultiControlOps
 2/20 Test #256: nvqpp_MultiControlOps .............   Passed    3.05 sec
      Start 257: nvqpp_ExpVals
 3/20 Test #257: nvqpp_ExpVals .....................   Passed    2.95 sec
      Start 258: nvqpp_MidCircuitMeasurements
 4/20 Test #258: nvqpp_MidCircuitMeasurements ......   Passed    2.98 sec
      Start 259: nvqpp_PhaseEstimation
 5/20 Test #259: nvqpp_PhaseEstimation .............   Passed    3.12 sec
      Start 260: nvqpp_Grover
 6/20 Test #260: nvqpp_Grover ......................   Passed    3.01 sec
      Start 261: nvqpp_QAOA
 7/20 Test #261: nvqpp_QAOA ........................   Passed    3.22 sec
      Start 262: nvqpp_VQEH2
 8/20 Test #262: nvqpp_VQEH2 .......................   Passed    5.46 sec
      Start 263: nvqpp_AmplitudeEstimation
 9/20 Test #263: nvqpp_AmplitudeEstimation .........   Passed    3.43 sec
      Start 264: nvqpp_Builder
10/20 Test #264: nvqpp_Builder .....................   Passed    3.29 sec
      Start 265: nvqpp_QAOABuilder
11/20 Test #265: nvqpp_QAOABuilder .................   Passed    3.44 sec
      Start 266: nvqpp_VQEH2Builder
12/20 Test #266: nvqpp_VQEH2Builder ................   Passed    7.21 sec
      Start 267: nvqpp_ComputeAction
13/20 Test #267: nvqpp_ComputeAction ...............   Passed    3.02 sec
      Start 268: nvqpp_Gradients
14/20 Test #268: nvqpp_Gradients ...................   Passed    3.14 sec
      Start 269: nvqpp_IterativePhaseEstimation
15/20 Test #269: nvqpp_IterativePhaseEstimation ....   Passed    2.90 sec
      Start 270: nvqpp_RandomWalkPhaseEstimation
16/20 Test #270: nvqpp_RandomWalkPhaseEstimation ...   Passed    2.78 sec
      Start 271: nvqpp_CuQuantumBackend
17/20 Test #271: nvqpp_CuQuantumBackend ............   Passed    5.01 sec
      Start 272: nvqpp_SampleAsync
18/20 Test #272: nvqpp_SampleAsync .................   Passed    6.11 sec
      Start 273: nvqpp_ObserveMQPU
19/20 Test #273: nvqpp_ObserveMQPU .................   Passed    6.16 sec
      Start 274: nvqpp_ObserveMQPU_MPI
20/20 Test #274: nvqpp_ObserveMQPU_MPI .............   Passed    7.67 sec
```

```
root@d48b0968fd6d:/cuda-quantum/build# ctest -R pycudaq_
Test project /cuda-quantum/build
    Start 275: pycudaq_Intro
1/8 Test #275: pycudaq_Intro ....................   Passed    0.15 sec
    Start 276: pycudaq_BernsteinVazirani
2/8 Test #276: pycudaq_BernsteinVazirani ........   Passed    0.16 sec
    Start 277: pycudaq_QAOA
3/8 Test #277: pycudaq_QAOA .....................   Passed    0.36 sec
    Start 278: pycudaq_VQE
4/8 Test #278: pycudaq_VQE ......................   Passed    0.16 sec
    Start 279: pycudaq_VQEAdvanced
5/8 Test #279: pycudaq_VQEAdvanced ..............   Passed    0.18 sec
    Start 280: pycudaq_SampleAsync
6/8 Test #280: pycudaq_SampleAsync ..............   Passed    2.93 sec
    Start 281: pycudaq_ObserveMQPU
7/8 Test #281: pycudaq_ObserveMQPU ..............   Passed    2.87 sec
    Start 282: pycudaq_ObserveMQPU_MPI
8/8 Test #282: pycudaq_ObserveMQPU_MPI ..........   Passed    4.43 sec
```

Reply:
Much better, thanks!